File: data/parking-rules-audit/verify-live.py
import json,httpx
from pathlib import Path
root=Path('data/parking-rules-audit');c=httpx.Client(base_url='http://127.0.0.1:8095',timeout=180)
def get(p):r=c.get('/api/'+p);r.raise_for_status();return r.json()
def command(action,p):return c.post('/api/command/'+action,json=p,headers={'X-Control-Client':'carla-control-center'})
s=get('status');cfg=get('configuration');old=json.loads((root/'restart-configuration.json').read_text());m=get('map')
assert s['recovery_operation']['stage']=='ready' and not s['running']
assert len(cfg['actors'])==len(old['actors'])==28 and len(s['sensors'])==6
assert [b['id'] for b in m['parking_spaces']]==json.loads((root/'report.json').read_text())['allowed_ids'];assert len(m['parking_excluded'])==28
old_ego=next(a for a in old['actors'] if a['role']=='ego');ego=next(a for a in cfg['actors'] if a['role']=='ego');assert ego['model']==old_ego['model']=='vehicle.lincoln.mkz_interior';assert sorted(ego['sensors'],key=lambda c:c['name'])==sorted(old_ego['sensors'],key=lambda c:c['name'])
r=command('spawn',{'role':'background','model':'vehicle.mini.cooper','parking_space':'P009'});assert r.status_code==400 and 'parking bay' in r.text,r.text
bg=next(a for a in cfg['actors'] if a['role']=='background');r=command('destination',{'id':bg['id'],'point':{'parking_space':'P009'}});assert r.status_code==400 and 'parking bay' in r.text,r.text
new=get('configuration');assert len(new['actors'])==28;assert next(a for a in new['actors'] if a['id']==bg['id'])['destination']==bg['destination']
r=command('run',{});r.raise_for_status();s=get('status');assert s['running'];assert not s.get('error'),s['error']
(root/'final-status.json').write_text(json.dumps(s,indent=2));(root/'final-configuration.json').write_text(json.dumps(get('configuration'),indent=2));report={'actors':28,'sensors':6,'ego':ego['id'],'running':True,'selectable':13,'excluded':28,'invalid_spawn_and_destination_rejected':True};(root/'live-verification.json').write_text(json.dumps(report,indent=2));print(report)

File: data/parking-retarget/verify-live.mjs
import fs from 'node:fs';import assert from 'node:assert/strict';import {chromium} from '@playwright/test';
const browser=await chromium.launch({headless:true,executablePath:'/home/william/.cache/ms-playwright/chromium-1217/chrome-linux64/chrome',args:['--no-sandbox']});
const page=await browser.newPage({viewport:{width:1440,height:1000}}),errors=[];page.on('pageerror',e=>errors.push(e.message));
const url='http://127.0.0.1:8095';let aid;const command=async(action,data={})=>{
 const r=await page.request.post(url+'/api/command/'+action,{data,headers:{'X-Control-Client':'carla-control-center'}});assert.ok(r.ok(),await r.text());return r.json();
};
try {
 let s=await (await page.request.get(url+'/api/status')).json();assert.equal(s.recovery_operation.stage,'ready');assert.equal(s.running,false);
 const before=JSON.parse(fs.readFileSync('data/parking-retarget/before-configuration.json'));
 const config=await (await page.request.get(url+'/api/configuration')).json();
 const oldEgo=before.actors.find(a=>a.role==='ego'),newEgo=config.actors.find(a=>a.role==='ego');
 assert.equal(newEgo.model,oldEgo.model);assert.deepEqual(newEgo.sensors,oldEgo.sensors);assert.equal(Object.keys(s.managed).length,28);
 const spawn=await command('spawn',{role:'background',model:'vehicle.mini.cooper',parking_space:'P025'});aid=spawn.id;
 await page.goto(url+'/');await page.waitForFunction(()=>document.querySelector('#actor-count').textContent==='29');
 const dismiss=page.getByRole('button',{name:'Continue to scene',exact:true});if(await dismiss.isVisible())await dismiss.click();
 await page.locator('nav button[data-panel="scenario"]').click();await page.locator('#actor-list button[data-id="'+aid+'"]').click();
 const updates=[];
 for(const option of ['68','132']){
  const response=page.waitForResponse(r=>r.url().endsWith('/api/command/destination')&&r.request().method()==='POST');
  await page.locator('#goal-select').selectOption(option);const r=await response;assert.ok(r.ok(),await r.text());const body=await r.json();updates.push(body);
  await page.waitForFunction(()=>!document.querySelector('#goal-select').disabled);
  assert.equal(body.parked,false);assert.equal(body.arrived,false);assert.equal(body.parking_trip.stage,'leaving');
 }
 assert.equal(updates[1].route_update.revision,updates[0].route_update.revision+1);
 assert.notDeepEqual(updates[0].destination,updates[1].destination);
 assert.match(await page.locator('#destination-status').textContent(),/Leaving parking/);
 s=await (await page.request.get(url+'/api/status')).json();const start=s.actors.find(a=>a.id===aid).pose;
 for(let i=0;i<12;i++)await command('step');
 s=await (await page.request.get(url+'/api/status')).json();const end=s.actors.find(a=>a.id===aid).pose;
 const distance=Math.hypot(end.x-start.x,end.y-start.y);assert.ok(distance>.02,'Parked actor did not begin moving');
 assert.equal(s.managed[aid].parking_trip.blocked,null);assert.deepEqual(errors,[]);
 await page.screenshot({path:'data/parking-retarget/deployed-interface.png',fullPage:true});
 const report={verified:true,temporary_actor:aid,paused_dropdown_changes:2,route_revisions:updates.map(u=>u.route_update.revision),stage:s.managed[aid].parking_trip.stage,distance_m:distance,ego_sensors_preserved:newEgo.sensors.length,errors};
 fs.writeFileSync('data/parking-retarget/live-verification.json',JSON.stringify(report,null,2));console.log(report);
}finally{
 if(aid!==undefined)await command('delete',{id:aid});
 await browser.close();
}
